src/main/kotlin/com/exactpro/th2/codec/xml/XmlPipelineCodec.kt

```diff
@@ -0,0 +1,570 @@
+
+import com.exactpro.sf.common.messages.structures.IDictionaryStructure
+import com.exactpro.sf.common.messages.structures.IFieldStructure
+import com.exactpro.sf.common.messages.structures.IMessageStructure
+import com.exactpro.sf.externalapi.codec.impl.EncodeException
+import com.exactpro.th2.codec.DecodeException
+import com.exactpro.th2.codec.api.IPipelineCodec
+import com.exactpro.th2.codec.api.IPipelineCodecSettings
+import com.exactpro.th2.common.grpc.AnyMessage
+import com.exactpro.th2.common.grpc.Direction
+import com.exactpro.th2.common.grpc.ListValue
+import com.exactpro.th2.common.grpc.Message
+import com.exactpro.th2.common.grpc.MessageGroup
+import com.exactpro.th2.common.grpc.RawMessage
+import com.exactpro.th2.common.grpc.Value
+import com.exactpro.th2.common.message.message
+import com.exactpro.th2.common.message.messageType
+import com.exactpro.th2.common.message.set
+import com.exactpro.th2.common.message.toJson
+import com.exactpro.th2.common.value.add
+import com.exactpro.th2.common.value.toValue
+import com.google.protobuf.ByteString
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import org.w3c.dom.Document
+import org.w3c.dom.Node
+import org.w3c.dom.NodeList
+import org.xml.sax.SAXException
+import java.io.ByteArrayInputStream
+import java.io.IOException
+import java.io.OutputStream
+import java.nio.charset.Charset
+import javax.xml.parsers.DocumentBuilder
+import javax.xml.parsers.DocumentBuilderFactory
+import javax.xml.parsers.ParserConfigurationException
+import javax.xml.transform.OutputKeys
+import javax.xml.transform.Transformer
+import javax.xml.transform.TransformerConfigurationException
+import javax.xml.transform.TransformerException
+import javax.xml.transform.TransformerFactory
+import javax.xml.transform.dom.DOMSource
+import javax.xml.transform.stream.StreamResult
+import javax.xml.xpath.XPath
+import javax.xml.xpath.XPathConstants
+import javax.xml.xpath.XPathFactory
+
+class XmlPipelineCodec : IPipelineCodec {
+
+    companion object {
+        const val XML_CHARSET_ATTRIBUTE = "XmlEncoding"
+        const val XML_DOCUMENT_TYPE_PUBLIC_FORMAT_STR_ATTRIBUTE = "XmlDocumentTypePublic"
+        const val XML_DOCUMENT_TYPE_SYSTEM_FORMAT_STR_ATTRIBUTE = "XmlDocumentTypeSystem"
+        const val XML_DOCUMENT_ROOT_TAG_ATTRIBUTE = "XmlRootTag"
+
+        const val XML_TAG_NAME_ATTRIBUTE = "XmlTagName"
+        const val XML_ATTRIBUTE_NAME_ATTRIBUTE = "XmlAttributeName"
+        const val XML_ATTRIBUTE_VALUE_ATTRIBUTE = "XmlAttributeValue"
+        const val EMBEDDED_ATTRIBUTE = "Embedded"
+
+        const val XML_X_PATH_EXPRESSION_ATTRIBUTE = "XPath"
+
+        private val LOGGER: Logger = LoggerFactory.getLogger(XmlPipelineCodec::class.java)
+        private val FORMAT_REPLACE_REGEX = Regex("\\{([0-9]+)}")
+        private val X_PATH: XPath = XPathFactory.newInstance().newXPath()
+        private val DOCUMENT_BUILDER: ThreadLocal<DocumentBuilder?> = ThreadLocal.withInitial {
```

Comment:
```suggestion
        private val DOCUMENT_BUILDER: ThreadLocal<DocumentBuilder> = ThreadLocal.withInitial {
```
Document builder shouldn't be nullable